packages/panzer/dof-mgr/src/Panzer_FieldAggPattern.cpp

```diff
@@ -180,45 +189,70 @@ void FieldAggPattern::buildFieldIdsVector()
 
    // numFields_: stores number of fields per geometric ID
    // fieldIds_:  stores field IDs for each entry field indicated by numFields_
-   numFields_.resize(geomAggPattern->numberIds());
+   numFields_.resize(geomAggPattern->numberIds(),0);
    fieldIds_.clear();
 
    // build numFields_ and fieldIds_ vectors
-   for(int dim=0;dim<getDimension()+1;dim++) {
-      int numSubcell = geomAggPattern->getSubcellCount(dim);
-      for(int sc=0;sc<numSubcell;sc++) {
-         // merge the field patterns for multiple fields
-         // on a specific dimension and subcell
-         mergeFieldPatterns(dim,sc);
-      }
-   }
+
+   // Merge the field patterns for multiple fields for each dimension
+   // and subcell. We do all the CG first, then all DG and finally all
+   // SKELETON. This allows us to use one offset for mapping DOFs to
+   // subcells later on.
+   this->mergeFieldPatterns(FieldType::CG);
+   this->mergeFieldPatterns(FieldType::DG);
+   this->mergeFieldPatterns(FieldType::SKELETON);
 }
 
-void FieldAggPattern::mergeFieldPatterns(int dim,int subcell)
+void FieldAggPattern::mergeFieldPatterns(const FieldType& fieldType)
 {
-   // make sure that the geometric IDs count is equal to the maxDOFs
-   const std::vector<int> & geomIndices = getGeometricAggFieldPattern()->getSubcellIndices(dim,subcell);
-
-   // a single geometric entity is assigned
-   // indices size should be either 1 or 0.
-   TEUCHOS_ASSERT(geomIndices.size()<=1);
-
-   if (geomIndices.size() > 0) {
-     const int geomIndex = geomIndices[0];
-     
-     numFields_[geomIndex] = 0; // no fields initially      
-
-     std::vector<std::pair<int,FPPtr> >::const_iterator itr;
-     for(itr=patterns_.begin();itr!=patterns_.end();++itr) {
-       std::size_t fieldSize = itr->second->getSubcellIndices(dim,subcell).size();
-       
-       // add field ID if their are enough in current pattern 
-       for (std::size_t i=0;i<fieldSize;++i)
-         fieldIds_.push_back(itr->first);
-       numFields_[geomIndex]+=fieldSize;
-     }
-     TEUCHOS_ASSERT(numFields_[geomIndex]>=0);
-   }
+  auto geomAggPattern = getGeometricAggFieldPattern();
+
+  // For DG/SKELETON, all DOFs are added to the internal cell
+  const int cellDim = this->getDimension();
+
+  // For CG and DG, add in all DOFs. For SKELETON do not add in
+  // interior cell DOFs. We can do this by reducing the dimension
+  // count by one.
+  int numDimensions = 0;
+  if ( (fieldType == FieldType::CG) || (fieldType == FieldType::DG) )
+    numDimensions = getDimension()+1;
+  else if (fieldType == FieldType::SKELETON)
+    numDimensions = getDimension();
+
+  for(int dim=0;dim<numDimensions;dim++) {
+    int numSubcell = geomAggPattern->getSubcellCount(dim);
+    for(int subcell=0;subcell<numSubcell;subcell++) {
+
+      // Get the geometric index to add the DOF indices to. CG adds to
+      // the subcell we are iterating over, (dim,subcel), while DG and
+      // SKELETON add to the internal cell (cellDim,0)
```

Comment:
Why does the skeleton add to the internal? They should be shared between cells for HDG.